Google-Summer-of-Code-2023.md

```diff
@@ -273,3 +273,504 @@ The following is a list of Oppia's 2023 GSoC project ideas. You are welcome to c
+
+**Useful resources:**
+
+### 4.4. Make CI faster
+
+**Project Description:**
+
+The broad aim of this project is to get the CI tests to run in under 1 hour so that PRs can be merged into develop quicker. This entails several steps:
+
+- Get the pre-push tests (that run on the developer’s local machine) to run in under 5 minutes:
+  - Audit backend unit tests to see which ones take a long time, and refactor those to avoid network/database/disk-access calls in order to shorten their runtime (while still testing what they need to). 
+  - For long backend unit tests that need to be long because they are creating a lot of data to test scalability, add them to a “scalability” list of tests that doesn’t need to get run during pre-push.
+  - Run only the lint, MyPy, and typescript checks; karma tests; and backend unit tests that were affected by files changed since the last push.
+  - (If possible) On pre-push, shorten the build step or remove it altogether.
+- Get the CI checks to run in under 1 hr wall time (i.e. including parallelization):
+  - Create a single environment at the start that gets used for all other tests. (In particular we currently seem to be running the “install third party deps” step repeatedly, so it might be good to cache folders like oppia_tools.)
+  - Create a build artifact that gets reused in all workflows. (Currently, build artifacts can’t be persisted across workflows, but this GitHub action for uploading to GCS might be helpful.)
+  - Run the full versions of all pre-push tests (mentioned above) first, and block other tests on these. (If any of the pre-push tests fail, then do not run the other CI checks.)
+  - For long-running tests:
+    - If the test is essential, find a way to make it shorter.
+    - If it is not, remove it from CI and put it in an acceptance layer that only runs on some release-candidate builds (see below). 
+- Split off less-important long-running tests into acceptance tests:
+- Create a GitHub Action that, at regular time intervals, takes the latest release-candidate build on develop that passes all CI checks, runs the acceptance tests on them. Builds that pass all acceptance tests should have a public, machine-readable indication that they did, and builds that cause an acceptance test failure should result in an email or chat notification to the CI Stability team.
+
+**Size of this project:** large (~350 hours)
+
+**Potential Mentors:** 
+
+**Knowledge/Skills Recommended:**
+
+**Suggested Milestones:**
+
+* **Milestone 1:** CI checks run in under an hour. Less important tests are split off into an acceptance layer.
+
+* **Milestone 2:** Pre-push checks run in under 5 minutes.
+
+**Dependency on Release Schedule:** 
+
+**Proposal notes:**
```

Comment:
I think it's worth linking to the idea doc (https://docs.google.com/document/d/1pIAu06LJS39BUeRyPZVWFYSW9aIKDrA_z7uoNtzoe9I/edit#) here since there's a lot more detail there and in the PRD